### ISSUE TITLE
Rework release to do external modularization in 3.1.0, internal modularization in 3.2.0

### ISSUE THREAD

**Issue by bvenners** [scalatest/scalatest#1692]
Hi Chee Seng,

Everyone I spoke with today thought 3.1.0/3.2.0 approach made sense. Please look at how you think we should best do this. Essentially this would just be the next two releases of ScalaTest. I'd like to make sure 3.2.0 works before releasing 3.1.0 final, so we should release a milestone release for 3.2.0 now as well as a milestone release of 3.1.0. Once these both look correct we can start the RC process for 3.1.0.

The approach is to have 3.1.0 be a stepping stone release that has external modularization and deprecations to prepare the way for future internal modularization. Then later have a 3.2.0 release that drops the internal deprecations and adds the internal modules. The 3.1.0 scalatest artifact would not include the third party integrations, so we would be modularizing externally in 3.1.0:

<img width="984" alt="Screen Shot 2019-09-04 at 11 26 04 AM" src="https://user-images.githubusercontent.com/120885/64298542-d9902c00-cf6d-11e9-8904-a8bdf91d1ed2.png">

We don't have to pull out scalatest-compatible yet, but I think we might as well. This would get Cats going. Any project can upgrade by 1) upgrading to 3.0.8 and clearing deprecation warnings (can use our scalafix tool), then 2) changing their version number from 3.0.8 to 3.1.0. At this point their tests should all compile and run as before, but they'll get a slew of deprecation warnings. If they apply our 3.1.0 scalafix tool, that will rewrite all of them. And again, all their tests will continue to run. In addition, if they were using any third party library, such as scalatestplus-scalacheck, they would need to add dependencies on those.

At that point they would be ready to upgrade to 3.2.0. Once we release 3.2.0, they can just change 3.1.0 in their build to 3.2.0 and everything would work like before. They will also need to bump the version number in their third party artifacts. They won't get any new deprecation warnings. All they would get is that behind the scenes, the scalatest artifact is just a pom referring to all the top-level internal modules, like this:

<img width="989" alt="Screen Shot 2019-09-04 at 11 24 07 AM" src="https://user-images.githubusercontent.com/120885/64298703-633ff980-cf6e-11e9-9c44-09a82c46e07f.png">

We would release a set of scalatestplus artifacts for 3.1.0 that depend on scalatest artifact, and another set of scalatestplus artifacts for 3.2.0 that depend on scalatest-core.

Please think about how best to reorganize the build to take this path.

Thanks.

Bill

**Comment by cheeseng** [comment]
@bvenners Sure that would work for me, eventually option 2 (the going to be 3.2.0) should be the goal we want to achieve.  Had a thought last night that perhaps 3.1.0 final should go out first and we start to test 3.2.0, honestly I don't think we have tested the modularization enough, so perhaps 3.2.0 shall go through some period of milestone/rc for testing also, at the same time we shall refactor our code base following the modules design and modularize their own tests also (one good tests for the module design, too).  The current module generation approach has the risk of things being missed out, e.g. one real example is that I missed the scalatest-funsuite module during the M1 release.

As first step I think we shall freeze 3.1.x (after RC1 PR that I am going to submit) and go ahead and release 3.1.0-RC1, then we shall branch off from 3.1.x to make 3.2.x-new (we already have 3.2.x branch that outdated).  We probably should still patch 3.1.x (should be stable for fat jar release as we have been doing so, but modules jar is untested to be honest).  I'll submit RC1 (fix for modularization) PR today.

**Comment by bvenners** [comment]
@cheeseng Yes but my concern is that if we find a problem with 3.2.0 after 3.1.0 is released final, we may need to but not be able to fix it in 3.1.0. What I was thinking is we should release 3.1.0-RC1 now, and also 3.2.0-M1 soon thereafter. I think we should make sure we get 3.2.0 correct by testing the milestone releases, and then when we are sure some 3.1.0-RCx is good and some 3.2.0-My is good, then we release 3.1.0 final. I.e., let's go ahead and do the work to do the modularization correctly in the project structure in a 3.2.x branch while 3.1.0 is in its RC phase.

Some time after we release 3.1.0 final we can start the 3.2.0-RC1 process. Historically we have always given folks one to two years after a deprecation before releasing a version that removes the deprecated item. For 3.0.8 and 3.1.0 we're going to do shorter deprecation cycles, but at least provide a ScalaFix tool for each.

**Comment by cheeseng** [comment]
@bvenners Sounds like a good plan to me, I'll start to work on 3.1.0-RC1, should be after I separate out scalatest-compatitble from scalatest.

Thanks!
